# Cleaning Scraped Used Car Data 





we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. The dataset was originally scraped and got cleaned. It then got messy. Our job is to re-clean the data. 


In [1]:
#Import the pandas and NumPy libraries
import pandas as pd
import numpy as np

#read in the autos.csv file into pd 
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [2]:
#let's check the data:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
# We will  print information about the autos dataframe, 
autos.info()
# as well as the first few rows
autos.head() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Note that the majority of cells have all 50,000 entries per column. However when we see numbers that are slightly less than 50,000 then this indicates there exists null values. We will see this more as we clean the data.

Also, most of the data types are objects. The implies that there could be multiple things in the cells labeled object. For example there can exist integers mixed with strings. When we see the objects that are specifically integers or floats then we know that the entire column has that data type .

***

We will first convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [4]:
# print columns and then change a few of the values to snakecase
print(autos.columns)


Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [5]:
autos.columns = ('date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen')


Check the first five rows to see if changes have been made:


In [6]:
print(autos.head())

          date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158 

***

 Let's take a look at the basic data and see what we need to manipulate.
 

In [7]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-16 21:50:53  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Most sellers are private sellers in the salon column. This is why there are only two unique values. Maybe this data is not needed.
 

On the offer type column there are only two unique values. The predominant one is a string named Angebot. 
This is the German word for services




A few other columns that need to be looked at our the following:
ab_test
registration_year	
gearbox	
power_ps
registration_month
ad_created	
num_photos	
postal_code

Whenever we see NA then that is a indication to take a closer look


 let's focus on the price and odometer columns:
 

We will remove non-numeric characters and convert it to a numeric data type: 
 

In [8]:
# four price we will remove all of the dollar signs $:
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")

print(autos["price"])

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: object


In [9]:
# note that it is still an object tired. We have to convert it to a numeric:
autos["price"] = autos["price"].astype(int)

#take a look and see how do you change the price data type to int
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  int64 
 5   ab_test             50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   power_ps            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer            50000 non-null  object
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

Now we will focus on the odometer column. We will do the same thing that we did for the price column. We will also change the column header name slightly:


In [10]:
# for Odometer we will remove all of the km Metrics and the commas:
autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].str.replace(",","")

# now we will change the data type to integer

# note that it is still an object tired. We have to convert it to a numeric:
autos["odometer"] = autos["odometer"].astype(int)

#take a look and see how do you change the price data type to int
autos.info()

# we will also change the name of the header:
autos = autos.rename(columns={"odometer":"odometer_km"})




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  int64 
 5   ab_test             50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   power_ps            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer            50000 non-null  int64 
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

In [11]:
autos.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   seller              50000 non-null  object
 3   offer_type          50000 non-null  object
 4   price               50000 non-null  int64 
 5   ab_test             50000 non-null  object
 6   vehicle_type        44905 non-null  object
 7   registration_year   50000 non-null  int64 
 8   gearbox             47320 non-null  object
 9   power_ps            50000 non-null  int64 
 10  model               47242 non-null  object
 11  odometer_km         50000 non-null  int64 
 12  registration_month  50000 non-null  int64 
 13  fuel_type           45518 non-null  object
 14  brand               50000 non-null  object
 15  unrepaired_damage   40171 non-null  object
 16  ad_created          50

***

In [12]:
#we explore the price column now: 

autos["price"].unique()

array([ 5000,  8500,  8990, ...,   385, 22200, 16995])

2,357 unbique values seems reasonable since people usually like to sell used cars at whole numbers. 

In [13]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Here the minimum value of zero dollars does not make sense. The maximum value of $100 million for a used car is also a bit much. Unless it is a very expensive car 

In [14]:
autos["price"].value_counts()

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64

Having 1421 items being zero dollars is a little bit alarming. There could be an error in this data. 

In [15]:
    autos["price"].value_counts().sort_index(ascending = True)

0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price, Length: 2357, dtype: int64

We will remove everything that costs zero dollars. We will also remove anything above a certain price. Let's take a look at the highest prices and see what we can delete:


In [16]:
autos["price"].value_counts().sort_index(ascending = False).head(30)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
194000      1
190000      1
180000      1
175000      1
169999      1
169000      1
163991      1
163500      1
155000      1
151990      1
Name: price, dtype: int64

 I am going to also delete entries that are above $190,000. This is arbitrarily packed. I just can't seem to reason that a lot of people would buy some thing that high priced on ebay.com
 

In [17]:
autos = autos[autos["price"].between(1,190000)]

In [18]:
#The remaining values seem reasonable
autos.describe()
 


price  registration_year      power_ps    odometer_km  \
count   48554.000000       48554.000000  48554.000000   48554.000000   
mean     5831.119105        2004.754665    117.143696  125791.386909   
std      8165.961802          88.653744    200.622842   39761.990894   
min         1.000000        1000.000000      0.000000    5000.000000   
25%      1200.000000        1999.000000     71.000000  125000.000000   
50%      3000.000000        2004.000000    107.000000  150000.000000   
75%      7450.000000        2008.000000    150.000000  150000.000000   
max    190000.000000        9999.000000  17700.000000  150000.000000   

       registration_month  num_photos   postal_code  
count         48554.00000     48554.0  48554.000000  
mean              5.78249         0.0  50972.124480  
std               3.68581         0.0  25748.017216  
min               0.00000         0.0   1067.000000  
25%               3.00000         0.0  30657.000000  
50%               6.00000         0.0  49716.000000  
75%               9.00000         0.0  71665.000000  
max              12.00000         0.0  99998.000000

Now let's take a look at the odometer column:

In [19]:
autos["odometer_km"].describe()

count     48554.000000
mean     125791.386909
std       39761.990894
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [20]:
autos["odometer_km"].value_counts().sort_index(ascending = True).head()

5000     830
10000    253
20000    760
30000    779
40000    815
Name: odometer_km, dtype: int64

The odometer values seem to be fine



***



In [21]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We will calculate the distribution of values for the three columns above. 

In [22]:
(autos['date_crawled']
        .str[:10]
        .value_counts(normalize = 1, dropna = False)
        .sort_index() # sorting by index ranks the date in ascending order
 
        )

# in order to read it properly, the .025333 number comes 
#from dividing the total amount of dates that are on March 5, 2016 
#divided by the total amount of crawling dates recorded



2016-03-05    0.025333
2016-03-06    0.014046
2016-03-07    0.036022
2016-03-08    0.033283
2016-03-09    0.033097
2016-03-10    0.032191
2016-03-11    0.032582
2016-03-12    0.036928
2016-03-13    0.015673
2016-03-14    0.036557
2016-03-15    0.034292
2016-03-16    0.029596
2016-03-17    0.031635
2016-03-18    0.012913
2016-03-19    0.034786
2016-03-20    0.037875
2016-03-21    0.037360
2016-03-22    0.032994
2016-03-23    0.032212
2016-03-24    0.029349
2016-03-25    0.031614
2016-03-26    0.032212
2016-03-27    0.031079
2016-03-28    0.034827
2016-03-29    0.034106
2016-03-30    0.033674
2016-03-31    0.031841
2016-04-01    0.033694
2016-04-02    0.035486
2016-04-03    0.038596
2016-04-04    0.036495
2016-04-05    0.013078
2016-04-06    0.003172
2016-04-07    0.001401
Name: date_crawled, dtype: float64

The percentages just show a small contribution to the overall fraction of a specific crawling date divided by the whole.

In [23]:
(autos['ad_created']
        .str[:10]
        .value_counts(normalize = True, dropna = False)
        .sort_index()
        )

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038843
2016-04-04    0.036866
2016-04-05    0.011801
2016-04-06    0.003254
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64

Here we see April 3 and April 4, 2016 have had more  entry is where the crawler spotted a new creation of advertisements.

In [24]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.001071
2016-03-06    0.004325
2016-03-07    0.005396
2016-03-08    0.007414
2016-03-09    0.009598
2016-03-10    0.010669
2016-03-11    0.012378
2016-03-12    0.023788
2016-03-13    0.008897
2016-03-14    0.012605
2016-03-15    0.015879
2016-03-16    0.016435
2016-03-17    0.028092
2016-03-18    0.007353
2016-03-19    0.015838
2016-03-20    0.020657
2016-03-21    0.020616
2016-03-22    0.021378
2016-03-23    0.018536
2016-03-24    0.019772
2016-03-25    0.019216
2016-03-26    0.016806
2016-03-27    0.015632
2016-03-28    0.020843
2016-03-29    0.022346
2016-03-30    0.024777
2016-03-31    0.023788
2016-04-01    0.022799
2016-04-02    0.024921
2016-04-03    0.025188
2016-04-04    0.024488
2016-04-05    0.124748
2016-04-06    0.221815
2016-04-07    0.131936
Name: last_seen, dtype: float64

Note that on the 6th and 7th of April 2016 there was a very large contribution of ads that were last scene.
 This probably is because the last day the crawler was crawling on the website was on the 7th April. We noticed because we don't see any further entries above that go into the 8th April
 
 

Let's take a look at the registration year and get some details on that. It is a numerical data type so it will be easy to describe it:

In [25]:
autos["registration_year"].describe()

count    48554.000000
mean      2004.754665
std         88.653744
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Since these values are supposed to represent years, the minimum and maximum values do not make sense. The minimum value of being the year of 1000 is impossible for a car to have been made then since they did not exist. Also the maximum year of 9999 is way far in the future and this implies an error in the data.

***

Because of the registration year issue that we just witnessed, we will take out some of the lowest and highest values in the registration year call in. This will help us have more accurate data. First let's say the number of listings that fall outside the span of 1900 to 2016:
 

In [26]:
autos["registration_year"].value_counts(normalize = 1).sort_index().head(20)

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000103
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000041
1937    0.000082
1938    0.000021
1939    0.000021
1941    0.000041
1943    0.000021
1948    0.000021
1950    0.000062
1951    0.000041
1952    0.000021
1953    0.000021
1954    0.000041
Name: registration_year, dtype: float64

In [27]:
autos["registration_year"].value_counts(normalize = 1).sort_index().tail(20)

2010    0.032726
2011    0.033427
2012    0.026960
2013    0.016538
2014    0.013655
2015    0.008012
2016    0.025065
2017    0.028669
2018    0.009680
2019    0.000041
2800    0.000021
4100    0.000021
4500    0.000021
4800    0.000021
5000    0.000082
5911    0.000021
6200    0.000021
8888    0.000021
9000    0.000021
9999    0.000062
Name: registration_year, dtype: float64

It is safe to assume that we should only have cars between 1910 and 2016 that logically makes sense. If in 1910 is a little old but it is still possible. And that is only possible because the gas engine car was created in 1886. Any year after that is fair game. And our candidate is 1910 and onwards. We will remove these values that don't make sense:
 

In [28]:
autos = autos[autos["registration_year"].between(1910,2016)]
autos["registration_year"].value_counts(normalize=True).sort_index()

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
          ...   
2012    0.028048
2013    0.017206
2014    0.014206
2015    0.008335
2016    0.026077
Name: registration_year, Length: 78, dtype: float64

***

Now we will look at the values in the brand column:
 

In [29]:
autos["brand"].value_counts(normalize=1)

volkswagen        0.211313
bmw               0.110028
opel              0.107607
mercedes_benz     0.096486
audi              0.086587
ford              0.069916
renault           0.047161
peugeot           0.029848
fiat              0.025648
seat              0.018277
skoda             0.016413
nissan            0.015277
mazda             0.015192
smart             0.014163
citroen           0.014013
toyota            0.012706
hyundai           0.010028
sonstige_autos    0.009771
volvo             0.009149
mini              0.008764
mitsubishi        0.008228
honda             0.007842
kia               0.007071
alfa_romeo        0.006642
porsche           0.005978
suzuki            0.005935
chevrolet         0.005700
chrysler          0.003514
dacia             0.002636
daihatsu          0.002507
jeep              0.002271
subaru            0.002143
land_rover        0.002100
saab              0.001650
jaguar            0.001564
daewoo            0.001500
trabant           0.001393
r

 we will only consider the brands of Volkswagen through Ford just because they make up the majority of listings. More than half… Below we were right code to  isolate the other brands and only work with the data from Ford and above:

In [30]:
brand_counts = autos["brand"].value_counts(normalize =1)
included_brands = brand_counts[brand_counts > .05].index #use .index to just show the row titles 

included_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

We will create an empty dictionary to hold aggregate data.
Then we'll Loop over included_brands, and assign the mean price to the dictionary, with the brand name as the key.

In [41]:
mean_price = {}
for brands in included_brands:
    brand_only = autos[autos["brand"]== brands]
    avg_price = brand_only["price"].mean()
    mean_price[brands] = int(avg_price)
mean_price
    

{'volkswagen': 5402,
 'bmw': 8236,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

It is clear that the most expensive cars are the German ones when considering the top six most common brands sold on eBay.
 

***

Our next goal is to use the aggregation method to compare the average mileage  and the mean price of the cars that we are focusing on. In order to do this we will combine data from both series objects and put them into a new data frame. The car brand names will be the new index of the state of frame. We will use series constructor and data frame constructor in order to build this.
 

In [32]:
# first we will create a dictionary for the average mileage below. 
#This is similar to the mean value dictionary


mean_mile = {}
for brands in included_brands:
    brand_only = autos[autos["brand"]== brands]
    avg_mile = brand_only["odometer_km"].mean()
    mean_mile[brands] = int(avg_mile)

mean_mile
    


{'volkswagen': 128707,
 'bmw': 132619,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

In [57]:
mean_mileage = pd.Series(mean_mile).sort_values(ascending=False)
mean_prices = pd.Series(mean_price).sort_values(ascending=False)

final_df = pd.DataFrame(mean_mileage, columns =['mean_milage'])
final_df



mean_milage
bmw                 132619
mercedes_benz       130788
opel                129310
audi                129157
volkswagen          128707
ford                124266

In [58]:
final_df['mean_price'] = mean_prices

In [59]:
final_df

mean_milage  mean_price
bmw                 132619        8236
mercedes_benz       130788        8628
opel                129310        2975
audi                129157        9336
volkswagen          128707        5402
ford                124266        3749

One can see that German cars, regardless of mileage are usually double the price of American made cars such as Opel or Ford. This could be due to the build quality. It is definitely better with German cars.  we can see that even if there's approximately the same amount of miles on a used German car versus a used American car, the German car was still cost substantially more on average.
 
  One must see a drastic increase in mileage on a German car for it to equate in price of an American car with substantially less mileage. Our data doesn't show this but if, for example, the BMW had around 200,000 miles, then we would probably see similar results in terms of average price being similar to the Ford which has 124,000 miles. 200,000 miles was arbitrarily picked. But the idea persists
  
  